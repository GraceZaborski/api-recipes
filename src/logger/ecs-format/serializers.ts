import { IncomingHttpHeaders } from 'http';

type ResponseSerializeObject = Record<string, unknown> & {
  statusCode?: number;
};

type RequestSerializeObject = Record<string, unknown> & {
  id: string;
  headers: IncomingHttpHeaders;
};

const SENSITIVE_HEADERS = [
  'x-token-payload',
  'authorization',
  'proxy-authorization',
  'proxy-authenticate',
  'www-authenticate',
];

const redactProps = <T extends Record<string, unknown>>(
  props: string[],
  data: T,
) =>
  props.reduce<T>(
    (res, key) => (res[key] ? { ...res, [key]: '[REDACTED]' } : res),
    data,
  );

function serializeResponse(object: ResponseSerializeObject) {
  const { statusCode, ...response } = object;

  return {
    ...response,
    status_code: statusCode,
  };
}

function serializeRequest(object: RequestSerializeObject) {
  const { id, ...request } = object;

  request.headers = redactProps(SENSITIVE_HEADERS, request.headers);

  return {
    ...request,
    id:
      request.headers['cf-request-id'] || request.headers['x-request-id'] || id,
    url: undefined,
  };
}

export const serializers = {
  'http.response': serializeResponse,
  'http.request': serializeRequest,
};
