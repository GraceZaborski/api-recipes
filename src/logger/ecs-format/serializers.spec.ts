import { serializers } from './serializers';

describe('http.response', () => {
  const responseSerializer = serializers['http.response'];

  it('should transform statusCode to status_code', () => {
    const mockResponse = {
      headers: {
        'content-type': 'application/json',
      },
      statusCode: 200,
    };

    expect(responseSerializer(mockResponse)).toMatchSnapshot();
  });
});

describe('http.request', () => {
  const requestSerializer = serializers['http.request'];

  it('should replace id with `cf-request-id`', () => {
    const mockRequest = {
      id: 'mock-id',
      url: '/mock-url',
      headers: {
        'cf-request-id': 'mock-id-cf',
      },
    };

    expect(requestSerializer(mockRequest)).toMatchSnapshot();
  });

  it('should replace id with `x-request-id`', () => {
    const mockRequest = {
      id: 'mock-id',
      url: '/mock-url',
      headers: {
        'x-request-id': 'mock-id-x',
      },
    };

    expect(requestSerializer(mockRequest)).toMatchSnapshot();
  });

  it('should fallback to id', () => {
    const mockRequest = {
      id: 'mock-id',
      url: '/mock-url',
      headers: {},
    };

    expect(requestSerializer(mockRequest)).toMatchSnapshot();
  });

  it('should redact sensitive headers', () => {
    const mockRequest = {
      id: 'mock-id',
      url: '/mock-url',
      headers: {
        'x-token-payload': 'sensitive-data',
        authorization: 'sensitive-data',
        'proxy-authorization': 'sensitive-data',
        'proxy-authenticate': 'sensitive-data',
        'www-authenticate': 'sensitive-data',
        'x-beamery-user-id': 'mock-user',
      },
    };

    expect(requestSerializer(mockRequest)).toMatchSnapshot();
  });
});
