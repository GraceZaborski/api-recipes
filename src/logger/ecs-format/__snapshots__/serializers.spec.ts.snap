// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`http.request should fallback to id 1`] = `
Object {
  "headers": Object {},
  "id": "mock-id",
  "url": undefined,
}
`;

exports[`http.request should redact sensitive headers 1`] = `
Object {
  "headers": Object {
    "authorization": "[REDACTED]",
    "proxy-authenticate": "[REDACTED]",
    "proxy-authorization": "[REDACTED]",
    "www-authenticate": "[REDACTED]",
    "x-beamery-user-id": "mock-user",
    "x-token-payload": "[REDACTED]",
  },
  "id": "mock-id",
  "url": undefined,
}
`;

exports[`http.request should replace id with \`cf-request-id\` 1`] = `
Object {
  "headers": Object {
    "cf-request-id": "mock-id-cf",
  },
  "id": "mock-id-cf",
  "url": undefined,
}
`;

exports[`http.request should replace id with \`x-request-id\` 1`] = `
Object {
  "headers": Object {
    "x-request-id": "mock-id-x",
  },
  "id": "mock-id-x",
  "url": undefined,
}
`;

exports[`http.response should transform statusCode to status_code 1`] = `
Object {
  "headers": Object {
    "content-type": "application/json",
  },
  "status_code": 200,
}
`;
